### ISSUE TITLE
Race condition after creating files

### ISSUE THREAD

**Issue by mertkahyaoglu** [jekyll/jekyll-admin#289]
Copying from #281 

I've encountered an issue while renaming files over and over again. Sometimes, after writing to the disc, the resulted page/document might be nil which causes 404. As far as I see, this happens randomly. Sometimes it renames and returns the renamed file, sometimes it returns nil.

IMO, this is because `site.process` is not actually blocking. Sometimes it takes so much time to finish so that there occurs a race condition which makes newly created files not ready to be served.

**Comment by mertkahyaoglu** [comment]
@parkr May I get your help here? How can we make `site.process` blocking [here](https://github.com/jekyll/jekyll-admin/blob/master/lib/jekyll-admin/server.rb#L72)?

**Comment by parkr** [comment]
Hey! So all of Jekyll is blocking, but when you run `jekyll serve`, we boot up two threads: one for the watcher (regenerates your site) and one for the HTTP server. It sounds like the presence of 2 threads is causing the confusion and overlap here.

To get around this, I would consider how you can achieve what you're looking for by hooking into the `site.process` method via `Jekyll::Hook`s: http://jekyllrb.com/docs/plugins/#hooks

**Comment by benbalter** [comment]
@parkr it doesn't look like there's a hook to know when the site's done building, is there?

**Comment by mertkahyaoglu** [comment]
@parkr @benbalter To avoid this overlap, can't we use some specific methods within `site.process` instead of calling all of them?

`site.process` consists of the following methods;

```ruby
def process
  self.reset
  self.read
  self.generate
  self.render
  self.cleanup
  self.write
end
````

What about just using `site.read` after writing to a file? Looks like doing this fixed the problem but I'm not sure this can cause any other problems.

**Comment by mertkahyaoglu** [comment]
Closed via #320

**Comment by mertkahyaoglu** [comment]
I'm trying to bring the `watch` back for Jekyll Admin and I think I found something to fix the race condition or (decrease the possibility to happen).

I discovered this race condition when I was trying to rename documents several times. I believe the reason is that we call `site.process` twice after a rename call. One for deleting a document and one for writing it to disk. 

After I removed the `site.process` call for deleting and just calling `site.read` after writing to disk, the race condition appears to be fixed. I'm testing this almost 💯  times and I haven't been able to get an error yet.

We can simply create a version of `delete_file` method which does not call `site.process` and use it for renaming purpose.

What do you guys think about this? Does it sound a solution for you?

**Comment by ashmaroli** [comment]
To add to this discussion,

I set up a new branch off current `master`, only commenting out `requiring` the monkey-patch file at `jekyll/build.rb` and started the development server.

Then proceeded with `creating`, `renamimg`, `deleting` of various `content_types` many times..

My results:
  - A Race condition (RC) occurs **everytime** I attempt to create a new `Page` / `Document` with just the title (path and body left blank)
   - A (RC) occasionally occurs when renaming a file path of `Page` / `Document` to a deeper directory
   - **Did not encounter a single occurrence of (RC) when creating / renaming / relocating / deleting a `DataFile` or when deleting a `Page` / `Document` **

(So.. whatever we're doing with Data files may be just about sufficient to eliminate this situation.
But unfortunately, that's not practical enough.)

**Comment by mertkahyaoglu** [comment]
@ashmaroli Did you try with the changes I introduced in the above comment ?

**Comment by ashmaroli** [comment]
@mertkahyaoglu I did try the changes in [`3b0a0ac`](https://github.com/jekyll/jekyll-admin/commit/3b0a0ac2bfad87b0dd4d967d93dd9d06d4c41931) in `development` mode only. 
Did not try with a `production` build

While in development mode, I did encounter Race Condition (RC) errors numerously while **creating a new page with just populating the `title` field**, and occasionally encountered RC while renaming.

```
/gems/jekyll-3.5.0/lib/jekyll/reader.rb:42: warning: conflicting chdir during another chdir block
```

**Comment by mertkahyaoglu** [comment]
Closed via #415
